tools/psa/__init__.py

```diff
@@ -15,31 +15,65 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from os.path import basename, splitext
+import os
+import shutil
+
 from tools.resources import FileType
+from tools.settings import ROOT
+from .generate_partition_code import manifests_discovery, generate_spm_code
+
 
 
-def find_secure_image(notify, resources, ns_image_path, configured_s_image_filename, image_type):
+def find_secure_image(notify, resources, ns_image_path,
+                      configured_s_image_filename, image_type):
     """ Find secure image. """
     if configured_s_image_filename is None:
         return None
 
-    assert ns_image_path and configured_s_image_filename, 'ns_image_path and configured_s_image_path are mandatory'
-    assert image_type in [FileType.BIN, FileType.HEX], 'image_type must be of type BIN or HEX'
+    assert ns_image_path and configured_s_image_filename, \
+        'ns_image_path and configured_s_image_path are mandatory'
+    assert image_type in [FileType.BIN, FileType.HEX], \
+        'image_type must be of type BIN or HEX'
 
     image_files = resources.get_file_paths(image_type)
     assert image_files, 'No image files found for this target'
 
-    secure_image = next((f for f in image_files if basename(f) == configured_s_image_filename), None)
     secure_image = next(
-        (f for f in image_files if splitext(basename(f))[0] == splitext(basename(ns_image_path))[0]),
+        (f for f in image_files if
+         os.path.basename(f) == configured_s_image_filename), None)
+    secure_image = next(
+        (f for f in image_files if
+         os.path.splitext(os.path.basename(f))[0] ==
+         os.path.splitext(os.path.basename(ns_image_path))[0]),
         secure_image
     )
 
     if secure_image:
         notify.debug("Secure image file found: %s." % secure_image)
     else:
-        notify.debug("Secure image file %s not found. Aborting." % configured_s_image_filename)
+        notify.debug("Secure image file %s not found. Aborting."
+                     % configured_s_image_filename)
         raise Exception("Required secure image not found.")
 
     return secure_image
+
+
+def _get_psa_autogen_dir():
+    return os.path.join(ROOT, 'PSA_AUTOGEN')
```

Comment:
after never-ending fights with build-system we can surely say, that the root folder of mbed-os is the most suitable place to store PSA auto-generated files (most of them are headers)

Reply:
I guess I don't know for sure, I'm just going off of this commit message: https://github.com/ARMmbed/mbed-os/pull/10447/commits/8526f7150e3ed6122ba35aed0132a264a47f8033

Reply:
the commit message is outdated

Reply:
I tried really hard to place the generated files under build folder but ran into a wall.
The way build system copies header files for test build just does not make it possible for the following reason:
- build directory we get from mbed-os contains toolchain path components.
- we cannot step out of build directory as once provided via command line results will be unpredictable
- when header are copied from the build for they are copied with the relative path (BUILD/<target>/<toolchain>
- Resource scan strips all the resource paths contain <toolchain> as these are considered as legacy build paths then test fails

If we really want these files to be generated under BUILD directory more changes will be required. These can be done later as an optimizations. We need this feature for customer in 5.12.3.

Reply:
Ok, I don't quite follow the points you made @alzix (might be because I haven't finished my morning coffee 😄). If you like we can discuss this offline to get a plan in place going forward.

Reply:
If the commit message is indeed outdated, could you please update it?

Reply:
done